wasm-chain-integration/src/lib.rs

```diff
@@ -337,9 +355,105 @@ impl<'a> HasCommon for ReceiveHost<'a> {
 
     fn param(&self) -> &[u8] { &self.param }
 
-    fn metadata(&self) -> &ChainMetadata { &self.receive_ctx.metadata }
+    fn metadata(&self) -> &Self::MetadataType { self.receive_ctx.metadata() }
+
+    fn policies(&self) -> ExecResult<&Self::PolicyType> { self.receive_ctx.sender_policies() }
+}
+
+pub trait HasInitContext {
+    type MetadataType: HasChainMetadata;
+    type PolicyBytesType: AsRef<[u8]>;
+    type PolicyType: SerialPolicies<Self::PolicyBytesType>;
+
+    fn metadata(&self) -> &Self::MetadataType;
+    fn init_origin(&self) -> ExecResult<&AccountAddress>;
+    fn sender_policies(&self) -> ExecResult<&Self::PolicyType>;
+}
+
+impl HasInitContext for InitContext<Vec<OwnedPolicy>> {
+    type MetadataType = ChainMetadata;
+    type PolicyBytesType = Vec<u8>;
+    type PolicyType = Vec<OwnedPolicy>;
+
+    fn metadata(&self) -> &Self::MetadataType { &self.metadata }
+
+    fn init_origin(&self) -> ExecResult<&AccountAddress> { Ok(&self.init_origin) }
+
+    fn sender_policies(&self) -> ExecResult<&Self::PolicyType> { Ok(&self.sender_policies) }
+}
+
+impl<'a> HasInitContext for InitContext<&'a [u8]> {
+    type MetadataType = ChainMetadata;
+    type PolicyBytesType = &'a [u8];
+    type PolicyType = &'a [u8];
+
+    fn metadata(&self) -> &Self::MetadataType { &self.metadata }
+
+    fn init_origin(&self) -> ExecResult<&AccountAddress> { Ok(&self.init_origin) }
+
+    fn sender_policies(&self) -> ExecResult<&Self::PolicyType> { Ok(&self.sender_policies) }
+}
+
+pub trait HasReceiveContext {
```

Comment:
And likewise this type. I think at leas I will probably forget the intent and reasons for this in a month or two.